### PR TITLE
Remove the zebra-consensus AddBlock order workarounds

### LINKED ISSUE #969: Remove the zebra-consensus AddBlock order workarounds
I'm seeing a bunch of "there should be no holes in the current chain" panics during syncing.

I think these errors happen because the `CheckpointVerifier` verifies blocks in order, but the blocks are asynchronously committed by the `ChainVerifier`. So the actual commits can get out of order.

There are a few ways to resolve this issue:
* the block and checkpoint verifiers add blocks to the state
* the `ChainVerifier` can wait for the previous block to be committed to the state, ideally via a future from the state service, or
* the state service can queue blocks until the previous block arrives.

In the state design RFC #902, we decided that:
* the state service can queue blocks until the previous block arrives.

TODO:
- [x] Stop committing blocks in the chain verifier
- [x] Commit blocks in the checkpoint verifier
  - [x] Pass the state to the checkpoint verifier
  - [x] Commit any verified blocks in the returned future
- [x] Commit blocks in the block verifier

Optional:
- [x] Commit blocks in the state's call, not in the returned future
- [x] Panic in the state as soon as the chain gets out of order, rather than waiting to panic in the next block locator request
  - [x] Temporarily panic if there is a duplicate block at a particular height - *implied by the chain order check*

### PR DESCRIPTION
**Pull request by teor2345**
Closes #969.

Panics on out of order blocks in the state - that will be resolved by #1011.

**Note -  this PR should be merged using a rebase.**

We want to revert the commit - `Temporarily ignore failing state tests` after #1011 is implemented.

### CODE REVIEW

File: zebra-consensus/src/checkpoint.rs
Comment (yaahc) [inline zebra-consensus/src/checkpoint.rs]: unless you're working with a `&mut Vec` here I don't think there's any reason to use drain, just iterating over the container will implicitly call `IntoIterator` which in the case of a Vec takes ownership of the values and consumes the container.

```suggestion
                for qblock in valid_blocks {
```

File: zebra-consensus/src/checkpoint.rs
Comment (yaahc) [inline zebra-consensus/src/checkpoint.rs]: :+1:

File: zebra-state/src/on_disk.rs
Comment (yaahc) [inline zebra-state/src/on_disk.rs:174]: :+1:

File: zebra-state/src/on_disk.rs
Comment (yaahc) [inline zebra-state/src/on_disk.rs:63]: :+1:

File: zebra-state/tests/basic.rs
Comment (yaahc) [inline zebra-state/tests/basic.rs:59]: do we have an out of order test?

**Review by yaahc — DISMISSED** [review]
lgtm, 1 minor comment

**Comment by hdevalence** [comment]
xref https://github.com/ZcashFoundation/zebra/pull/902#issuecomment-688616536

File: zebra-state/tests/basic.rs
Comment (teor2345) [reply zebra-state/tests/basic.rs:59]: We can't enable it by default until #1011 is merged, but you're right, we should have an out of order test.

(Before this PR, all the tests were out of order.)

File: zebra-consensus/src/chain.rs
Comment (teor2345) [inline zebra-consensus/src/chain.rs]:
```suggestion
        // users, so we don't need to call the verifier's `poll_ready` here.
```

File: zebra-consensus/src/checkpoint.rs
Comment (teor2345) [reply zebra-consensus/src/checkpoint.rs]: I have become too reliant on clippy for hints like this. 🤷

File: zebra-state/tests/basic.rs
Comment (teor2345) [reply zebra-state/tests/basic.rs:59]: The CheckpointVerifier `continuous_blockchain_test` is unordered, which is why I disabled it:
https://github.com/ZcashFoundation/zebra/pull/1012/files#diff-4b3b16117479af65b38872f84005c2e1R213

But we should have tests that are guaranteed out of order, rather than just random execution order.

Comment (teor2345) [reply zebra-state/tests/basic.rs:59]: Added those tests in 2736ec7, then commented out the failing out-of-order tests in 533a401.

I also squashed the fixes into the branch and force-pushed, so we can merge this branch using a rebase. That way, we can revert 533a401 once we fix #1011.

3a51d73 is the squashed commit, the others are the same.

The old branch is here:
https://github.com/teor2345/zebra/tree/state-order-2020-09-07

Comment (yaahc) [reply zebra-state/tests/basic.rs:59]: thank you for the old branch ref, for anyone else who needs the incremental heres the link to the compare

https://github.com/teor2345/zebra/compare/state-order-2020-09-07...teor2345:state-order

edit: actually, this includes more changes than I expected, so i think this isn't precisely the diff of changes since my last review :/ _shakes fist at force push and history modification during PR review_

**Review by yaahc — DISMISSED** [review]
lgtm

File: zebra-consensus/src/chain.rs
Comment (hdevalence) [inline zebra-consensus/src/chain.rs:37]: Non-blocking on this PR:

IIRC this logic was added to help debug sync issues.  Do we still need it, if we sync correctly and can handle out of order blocks?  If not, could we simplify the chain verifier by removing it?

File: zebra-consensus/src/chain.rs
Comment (hdevalence) [inline zebra-consensus/src/chain.rs:117]: Should we remove this TODO? It seemed like the conclusion was that Tower's middleware is mostly standardized on `Box<dyn Error + Send + Sync + 'static>`, and I think a previous attempt to use `Report` got stuck on generic bounds.  If we don't have an immediate plan to change the error types, it might be good to remove the TODOs.  We can always work through the code and change them if the Tower situation changes.

File: zebra-consensus/src/chain.rs
Comment (hdevalence) [inline zebra-consensus/src/chain.rs]: did something go wrong with rustfmt here?

File: zebra-consensus/src/checkpoint.rs
Comment (hdevalence) [inline zebra-consensus/src/checkpoint.rs]: My preference would be to not change our code to optimize for this regression -- there's much more engineering effort applied to Rust than to our project, so my expectation is that the regression will be fixed and we'll then have to update our code, or more likely, we'll forget or be busy and end up with an unneeded optimization.

I think optimizing for this kind of thing only makes sense when (1) the code is functionally complete and (2) we have some measurement that it's a real performance impact.

File: zebra-consensus/src/chain.rs
Comment (yaahc) [reply zebra-consensus/src/chain.rs:117]: yea we can probably remove these TODOs. The particular issue that prevents this from being possible is that `Buffer` and `Batch` both needs error types that are `Clone`, which is just not possible to implement for Report. What we probably want to do instead is switch all of these service error types to use `TracedError` so that we can propagate spantraces back out of the service which then get used in the report. We don't quite get to put suggestions and what not in our reports from services but thats probably fine, they probably don't need to be where we do higher level diagnostics for errors.

File: zebra-consensus/src/checkpoint.rs
Comment (hdevalence) [inline zebra-consensus/src/checkpoint.rs]: I'm not sure this logic makes sense... the `result` that's sent to the caller is not the result of that caller's block, but the result of some other blocks.

But because we're allowed to commit blocks to the state out of order, I don't think it's necessary to have two codepaths.  The old code did
```rust
        async move {
            // Remove the Result<..., RecvError> wrapper from the channel future
            rx.await
                .expect("CheckpointVerifier does not leave dangling receivers")
        }
        .boxed()
```
and the new code can commit the block after it learns it's ready:
```rust
        async move {
            match rx
                .await
                .expect("CheckpointVerifier does not leave dangling receivers")
            {
                Ok(hash) => {
                    state_service
                        .oneshot(zebra_state::Request::AddBlock { block })
                        .await?;
                    Ok(hash)
                }
                Err(e) => Err(e),
            }
        }
        .boxed()
```

File: zebra-state/src/on_disk.rs
Comment (hdevalence) [inline zebra-state/src/on_disk.rs:82]: This seems fine, but I don't think it corresponds to the procedure described in the state RFC -- I think I don't understand the intended scope of this PR?

**Comment by hdevalence** [comment]
I got a little confused by the title of the PR, so I want to make sure I understand the scope of the PR.  The state design RFC makes a few different categories of changes:

1. Changes to the request / response API, creating `CommitBlock` and `CommitFinalizedBlock` and removing `AddBlock`;
2. Changes to the ordering semantics of block insertions, making it the state service's job to handle out of order `Commit*` requests;
3. Changes to the sled structures used to store data (which imply changes to how blocks are actually written into sled);
4. Addition of an in-memory store for non-finalized state (cannot be implemented, as-yet unspecified).

Which of these is this PR supposed to implement?  It looks like it does (2), per the changes to the checkpoint verifier, but there are also a bunch of changes to the state itself that I'm not sure how to categorize.

**Comment by yaahc** [comment]
@hdevalence 2 is the goal of this PR. The idea being that it makes https://github.com/ZcashFoundation/zebra/pull/1018 fail faster which will make it easier to test the subsequent changes from the State RFC to make sure we're correctly handling out of order blocks.

File: zebra-state/src/on_disk.rs
Comment (hdevalence) [reply zebra-state/src/on_disk.rs:82]: clarified here: https://github.com/ZcashFoundation/zebra/pull/1012#issuecomment-689131742

seems fine to do this, although we'll have to change it later, self-resolving this question.

File: zebra-consensus/src/chain.rs
Comment (hdevalence) [reply zebra-consensus/src/chain.rs:117]: Got it.  For now, I think it's probably easiest to keep using `BoxError`, because we know there's no problems, even if it doesn't retain all the context we'd like, and experiment with using `TracedError` after we ship a first version of Zebra.

File: zebra-consensus/src/checkpoint.rs
Comment (hdevalence) [reply zebra-consensus/src/checkpoint.rs]: Would be avoided by the suggestion in https://github.com/ZcashFoundation/zebra/pull/1012#discussion_r485181158

File: zebra-consensus/src/chain.rs
Comment (teor2345) [reply zebra-consensus/src/chain.rs:37]: I think we should remove the out of order debug logs after:
* the state service handles out of order blocks #1011
* the inbound service handles out of order gossiped blocks #889

I added this task to the checklist in the sync correctness tracking issue #884.

File: zebra-consensus/src/chain.rs
Comment (teor2345) [reply zebra-consensus/src/chain.rs:117]: I opened #1034 to make this change later.

I'm concerned that it will modify a few different files, and this PR already has conflicts with other PRs.

File: zebra-consensus/src/chain.rs
Comment (teor2345) [reply zebra-consensus/src/chain.rs]: Weird that it passed my rustfmt, and also our rustfmt checks.

File: zebra-consensus/src/checkpoint.rs
Comment (teor2345) [reply zebra-consensus/src/checkpoint.rs]: > I'm not sure this logic makes sense... the `result` that's sent to the caller is not the result of that caller's block, but the result of some other blocks.

Can you explain what you mean here?

As far as I can see, each `result` is based on a `qblock.block`, and it is sent to the corresponding `qblock.tx`.

The `rx` later in the async block is totally unrelated, although it might match one of the earlier `qblock.tx` channels.

But I can see an issue with error handling here, where the `?` operator takes errors from any qblock, and propagates them to the unrelated current block.

So I'll implement this change to fix that issue.

File: zebra-consensus/src/chain.rs
Comment (teor2345) [reply zebra-consensus/src/chain.rs]: Fixed and squashed into f9088027f207577b08b0a7450d1ce05bc2e31c54.

File: zebra-consensus/src/checkpoint.rs
Comment (teor2345) [reply zebra-consensus/src/checkpoint.rs]: Implemented and squashed into f9088027f207577b08b0a7450d1ce05bc2e31c54.

This change commits blocks in an even more unordered fashion, so I needed to ignore another test. Squashed into c042555ac714819b03efbc565bc0223fedf43102.

**Comment by teor2345** [comment]
> Which of these is this PR supposed to implement? It looks like it does (2), per the changes to the checkpoint verifier, but there are also a bunch of changes to the state itself that I'm not sure how to categorize.

Sorry this PR was confusing, I named it while we were still working out which changes depended on other changes. (And who would do them.)

I've updated the name so it's focused on the final purpose of the PR: removing AddBlock order workarounds, in preparation for the state service handling order in #1011.

Here's how some other changes made it into this PR:
* f3b817440c2bbe56de3412af09ffb2ce66dc3ccd perform serialised block writes in the state: I thought the PR would be able to modify more of the state when I made this change
* d14be33ebcca75e468b22c728da91854d274d52f Panic when state blocks are inserted out of order: in practice, *not* doing this check was a workaround for out of order blocks - instead, we would panic later, if the missing blocks lost the race with a query
* 5631bc62c713b974dcf97c5a9f337378a7263c2c Add tests for in-order state AddBlock requests: previously, the tests were carefully designed to avoid panics due to query races with state invariants - now we actually want to test the invariants
* c042555ac714819b03efbc565bc0223fedf43102 Temporarily ignore failing state tests - but we don't actually want to test the invariants until #1011 is completed

**Comment by teor2345** [comment]
I squashed today's fixes into the branch and force-pushed, so we can merge this branch using a rebase. That way, we can revert the `Temporarily ignore failing state tests` commit once we fix #1011.

f9088027f207577b08b0a7450d1ce05bc2e31c54 and c042555ac714819b03efbc565bc0223fedf43102 are modified commits, the others are the same.

The old branch is here:
https://github.com/teor2345/zebra/tree/state-order-2020-09-08

**Comment by teor2345** [comment]
I just merged #1009, so this PR might have merge conflicts.

**Review by hdevalence — APPROVED** [review]

**Pull request merged**
